receiver/dockerstatsreceiver/docker.go

```diff
@@ -0,0 +1,247 @@
+	dc.logger.Debug("Fetching container stats.", zap.String("id", container.ID))
+	statsCtx, cancel := context.WithTimeout(ctx, dc.config.Timeout)
+	containerStats, err := dc.client.ContainerStats(statsCtx, container.ID, false)
+	defer cancel()
+	if err != nil {
+		if docker.IsErrNotFound(err) {
+			dc.logger.Debug(
+				"Daemon reported container doesn't exist. Will no longer monitor.",
+				zap.String("id", container.ID),
+			)
+			dc.removeContainer(container.ID)
+		} else {
+			dc.logger.Warn(
+				"Could not fetch docker containerStats for container",
+				zap.String("id", container.ID),
+				zap.Error(err),
+			)
+		}
+
+		results <- Result{nil, err}
+		return
+	}
+
+	statsJSON, err := dc.toStatsJSON(containerStats, &container, results)
+	if err != nil { // results have been sent in converter
+		return
+	}
+
+	md, err := ContainerStatsToMetrics(statsJSON, &container, dc.config)
+	if err != nil {
+		dc.logger.Error(
+			"Could not convert docker containerStats for container id",
+			zap.String("id", container.ID),
+			zap.Error(err),
+		)
+		results <- Result{nil, err}
+		return
+	}
+
+	results <- Result{md, nil}
+}
+
+func (dc *DockerClient) toStatsJSON(
+	containerStats dtypes.ContainerStats,
+	container *DockerContainer,
+	results chan<- Result,
+) (*dtypes.StatsJSON, error) {
+	var statsJSON dtypes.StatsJSON
+	err := json.NewDecoder(containerStats.Body).Decode(&statsJSON)
+	containerStats.Body.Close()
+	if err != nil {
+		// EOF means there aren't any containerStats, perhaps because the container has been removed.
+		if err == io.EOF {
+			// It isn't indicative of actual error.
+			results <- Result{nil, nil}
+			return nil, err
+		}
+		dc.logger.Error(
+			"Could not parse docker containerStats for container id",
+			zap.String("id", container.ID),
+			zap.Error(err),
+		)
+		results <- Result{nil, err}
+		return nil, err
+	}
+	return &statsJSON, nil
+}
+
+// Queries inspect api and returns *ContainerJSON and true when container should be queried for stats,
+// nil and false otherwise.
+func (dc *DockerClient) inspectedContainerIsOfInterest(ctx context.Context, cid string) (*dtypes.ContainerJSON, bool) {
+	inspectCtx, cancel := context.WithTimeout(ctx, dc.config.Timeout)
+	container, err := dc.client.ContainerInspect(inspectCtx, cid)
+	defer cancel()
+	if err != nil {
+		dc.logger.Error(
+			"Could not inspect updated container",
+			zap.String("id", cid),
+			zap.Error(err),
+		)
+	} else if !dc.shouldBeExcluded(container.Config.Image) {
+		return &container, true
+	}
+	return nil, false
+}
+
+func (dc *DockerClient) persistContainer(containerJSON *dtypes.ContainerJSON) {
+	if containerJSON == nil {
+		return
+	}
+	cid := containerJSON.ID
+	dc.logger.Debug("Monitoring Docker container", zap.String("id", cid))
+	dc.containers[cid] = DockerContainer{
+		ContainerJSON: containerJSON,
+		EnvMap:        containerEnvToMap(containerJSON.Config.Env),
+	}
+}
+
+func (dc *DockerClient) removeContainer(cid string) {
+	delete(dc.containers, cid)
```

Comment:
It looks like `dc.containers` is read from and written to concurrently. Access should probably be synchronized.

Reply:
My personal opinion, though, is that synchronization should be avoided altogether if possible. What do you think about these two other options:
1) Keep the first implementation really simple (no, or minimal parallelism until it becomes evident that it's necessary)
2) Mutate long-lived structs from one, main goroutine, and restrict additional goroutines to just locally scoped data, which you can then save/munge/whatever in the main goroutine after their waitgroup is done.

Reply:
There was no planned concurrency in this first stage, but you're right this could be an issue if runner intervals are trivially short (not sure what else you're referring to?).  There will definitely need to be synchronization when the docker event handler loop is added in the following PR.  My original implementation before reducing the functionality effectively did 2), and claimed a lock on `Containers()` calls before returning the copied list to avoid using shared structures in the Receiver.

Reply:
Hey Ryan. I was referring to `FetchContainerStatsAndConvertToMetrics` which it looks like runs concurrently via
```
	for _, container := range containers {
		go func(dc DockerContainer) {
			r.client.FetchContainerStatsAndConvertToMetrics(r.runnerCtx, dc, results)
			wg.Done()
		}(container)
	}
```
So the fetch calls are all launched at around the same time but each of them can mutate `dc.containers` via `removeContainer` as a kind of a side effect. Maybe it would be good to either (in effect) synchronize `removeContainer` or return the container-to-remove and do that synchronously?

BTW I still like the idea of no/minimal concurrency as a first pass, but maybe others have a different perspective on that.

Reply:
I added a simple lock which will be necessary w/ later functionality and have chosen to avoid a synchronous cleanup that wouldn't be used in the long term.

Reply:
Yeah you are absolutely going to need concurrency given how slow the stats endpoint responds.

Reply:
Hey Ryan, did the simple lock get added to this PR? It looks like `dc.containers` is accessed/mutated concurrently here. Happy to jump on a chat tomorrow to discuss if you want. :-)